### PR TITLE
Refactor BindingCollection to replace ArrayList with List<Binding>

### PR DESCRIPTION
**Pull request by Jericho**
Related: #8140

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8364)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/BindingsCollection.cs [hunk 1]
```diff
@@ -15,25 +15,25 @@ namespace System.Windows.Forms
     [DefaultEvent(nameof(CollectionChanged))]
     public class BindingsCollection : BaseCollection
     {
-        private ArrayList _list;
+        private List<Binding> _list;
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/BindingsCollection.cs:18]: Let's go ahead and enable nullability here and change this to `List<Binding>?`

**Review by lonitra — COMMENTED** [review]
Just a comment about enabling nullability since we seem to be doing those checks in the file anyways.

File: src/System.Windows.Forms/src/System/Windows/Forms/BindingsCollection.cs (hunk 1 above)
Comment (Jericho) [reply src/System.Windows.Forms/src/System/Windows/Forms/BindingsCollection.cs:18]: sure.

File: src/System.Windows.Forms/src/System/Windows/Forms/BindingsCollection.cs [hunk 2]
```diff
@@ -15,25 +13,25 @@ namespace System.Windows.Forms
     [DefaultEvent(nameof(CollectionChanged))]
     public class BindingsCollection : BaseCollection
     {
-        private ArrayList _list;
-        private CollectionChangeEventHandler _onCollectionChanging;
-        private CollectionChangeEventHandler _onCollectionChanged;
+        private List<Binding>? _list;
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/BindingsCollection.cs:16]: This could be assigned as `private List<Binding> _list = new();` and remove nullable from it.

Comment (Jericho) [reply src/System.Windows.Forms/src/System/Windows/Forms/BindingsCollection.cs:16]: The change you are proposing would allow removing a lot of null checking code (for example: `_list ??= new()`). However, it seems to me there was a deliberate decision to delay private members instantiation as much as possible throughout this project (presumably to ensure instantiation of the class in question is as quick as possible) so I'm reluctant to make this change here unless you tell me that's exactly what you want.

Here are a few examples taken at random from other classes where nullability has been enabled but private member instantiation has not been changed like you are suggesting: [here](https://github.com/dotnet/winforms/blob/079b548eea0799bd64f21ecf93771e7e317a5cb5/src/System.Windows.Forms/src/System/Windows/Forms/Application.ThreadContext.cs#L52), [here](https://github.com/dotnet/winforms/blob/079b548eea0799bd64f21ecf93771e7e317a5cb5/src/System.Windows.Forms/src/System/Windows/Forms/Form.EnumThreadWindowsCallback.cs#L15), [here](https://github.com/dotnet/winforms/blob/079b548eea0799bd64f21ecf93771e7e317a5cb5/src/System.Windows.Forms/src/System/Windows/Forms/KeysConverter.cs#L21), [here](https://github.com/dotnet/winforms/blob/079b548eea0799bd64f21ecf93771e7e317a5cb5/src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs#L172) and [here](https://github.com/dotnet/winforms/blob/079b548eea0799bd64f21ecf93771e7e317a5cb5/src/System.Windows.Forms/src/System/Windows/Forms/TabPage.cs#L31).

Playing devil's advocate, I was able to find a handful of examples where private members instantiation is not delayed: [here](https://github.com/dotnet/winforms/blob/079b548eea0799bd64f21ecf93771e7e317a5cb5/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs#L49) and [here](https://github.com/dotnet/winforms/blob/079b548eea0799bd64f21ecf93771e7e317a5cb5/src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs#L138). 

The bottom line is that I can make the change you mentioned but let me know if that's what you want (in your comment, you said this `could` be done so I wasn't sure if that meant you wanted me to make the change or not).

Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/BindingsCollection.cs:16]: There is a difference between this initialization and some of the others you pointed out. In this case, any instance accessing list is either allocating it or returning `0` count. Same is not true for others ( ex : message filters, in which you do not allocate always when you access it and there are few code paths that made sense to check for null). 

>you said this could be done so I wasn't sure if that meant you wanted me to make the change or not

Given i was reviewing in the browser, wasn't sure if any compile-time/runtime errors prevented you from doing it. So wanted you to double check if it is safe to do so.

File: src/System.Windows.Forms/src/System/Windows/Forms/BindingsCollection.cs [hunk 3]
```diff
@@ -15,25 +13,25 @@ namespace System.Windows.Forms
...
         }
 
-        public override int Count => _list is null ? 0 : base.Count;
+        public override int Count => _list.Count;
 
         /// <summary>
         ///  Gets the bindings in the collection as an object.
         /// </summary>
-        protected override ArrayList List => _list ??= new ArrayList();
+        protected override ArrayList List => ArrayList.Adapter(_list);
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/BindingsCollection.cs:29]: is this breaking? Previously ArrayList and List would return the same instance, now a new instance is created each time

**Comment by elachlan** [comment]
bumping this so it doesn't get closed.

**Comment by elachlan** [comment]
@Jericho if its okay with you, I can possibly take over this?

**Comment by Jericho** [comment]
@elachlan sure but all feedback was addressed and just waiting for merge.

**Comment by elachlan** [comment]
I'll leave it to the team to finish the review.

**Comment by elachlan** [comment]
@dreddy-work I think this might be waiting on the team. Could someone please review it and see what needs to be done?

File: src/System.Windows.Forms/src/System/Windows/Forms/BindingsCollection.cs (hunk 3 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/BindingsCollection.cs:29]: @hughbe, It doesn't creates new list but just wraps around it.

**Comment by dreddy-work** [comment]
> @dreddy-work I think this might be waiting on the team. Could someone please review it and see what needs to be done?

I looked into it again and i feel my comment above about initializing list still make sense in this scenario. If @Jericho / @elachlan push a change with respect to that, I will merge it. Rest all look good to me.

**Comment by Jericho** [comment]
@dreddy-work your feedback was addressed on Dec 13 2022 in commit 8a7327d0feabc8c4d4089e8e194027cd528422e7

**Comment by dreddy-work** [comment]
> @dreddy-work your feedback was addressed on Dec 13 2022 in commit [8a7327d](https://github.com/dotnet/winforms/commit/8a7327d0feabc8c4d4089e8e194027cd528422e7)

Thanks @Jericho, I saw comment was not resolved but didn't cross check the commit. Thank you.

**Review by dreddy-work — DISMISSED** [review]

**Comment by dreddy-work** [comment]
@Jericho , can you rebase this to get it merged?

**Comment by dreddy-work** [comment]
src\System.Windows.Forms\src\System\Windows\Forms\BindingsCollection.cs(67,52): error RS0036: (NETCORE_ENGINEERING_TELEMETRY=Build) Symbol 'CollectionChanged' is missing nullability annotations in the declared API.

**Review by dreddy-work — APPROVED** [review]
